### ISSUE TITLE
Really Interesting Project

### ISSUE THREAD

**Issue by cbluth** [a-shine/butter#4]
Ive been reading your docs, and it looks like youve done a lot of research into decentralized and distributed systems.

this project is interesting, and i am curious how it works.

can you explain some of the basic mechanisms from this project, for example like transport, discovery, and lookup?

also, if you need help on certain parts of the project, i may be interested in digging into the code and contributing too

**Comment by a-shine** [comment]
Hi there,

I'm really glad to hear :) This project has been so interesting to work on
and I've really fallen in love with peer-to-peer system design as a result.
The project is really focused on unstructured peer-to-peer architectures
which presents some interesting challenges.

I'm in the midst of cleaning everything up and writing up my full
dissertation report which will describe everything in detail. In the
meantime, I'm more than happy to jump on a quick call to talk you through
the system and discuss some future directions. Else I'll be happy to send
the full report to you when complete. What works best for you?


Regards,

Alex

**Comment by cbluth** [comment]
@a-shine 

i have been doing similar research on p2p and distributed systems, and your project `butter` and your research is aligned with my interests. i have been testing my own code by connecting to bt dht via bep specs, and testing other mechanisms like discovery, and rendezvous, and nat/turn/stun/ice, i2p/tor, and am interested in pooling our efforts.

i am looking for someone or a project that i can collaborate with, so that i can contribute ideas/efforts towards the next-gen of p2p apps.

**Comment by a-shine** [comment]
Yeah, that sounds great! The next stage for Butter would be to develop a proper robust testbed for peer-to-peer systems that can emulate all sorts of network topologies and I would really like to implement a few more interesting information retrieval algorithms. Particularly, ISM (described in [this paper](https://www.researchgate.net/publication/3422686_Information_retrieval_techniques_for_peer-to-peer_networks)). I would really appreciate some help on wider internet discovery (especially since you seems to know quite a lot about it). It would be interesting to re-design a wider discovery module that allows peers to discover each other over the internet (hence overcoming NAT and then discovering each other). I think a good approach specially with something like NAT traversal is to implement a default solution e.g. using UPnP and then several fallback options if the router of the user doesn't have it enables - what do you think?
